examples/user-data-store/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+
+#[derive(uniffi::Error)]
+pub enum Error {
+    DatabaseError(String),
+    UnexpectedError(String),
+}
+
+impl From<uniffi::UnexpectedCallbackError> for Error {
+    fn from(_: uniffi::UnexpectedCallbackError) -> Self {
+        Self::UnexpectedError("Error invoking callback interface")
+    }
+}
+
+pub type Result<T> = Result<T, Error>;
+
+#[uniffi::callback_interface]
+pub trait Logger {
+    /// Log a message
+    async fn log(&self, message: String) -> CallbackResult<()>;
+}
+
+#[uniffi::callback_interface]
+pub trait CryptoKeyStore {
+    /// Get a key for cryptography operations.  Depending on the application, this may load the key
+    /// from a file or from an OS key store and may require the user to take some action to unloke
+    /// the key.
+    async fn get(&self) -> CallbackResult<String>;
+}
+
+/// Event notification system.
+#[uniffi::callback_interface]
+pub trait EventListener {
+    async fn db_operation_complete(&self, store: Arc<UserStore>, success: bool) -> CallbackResult<()>;
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct UserStore {
+    logger: Box<dyn Logger>,
+    key_store: Box<dyn KeyStore>,
+    listener: Box<dyn EventListener>,
+    queue: ThreadQueue,
+    // Use OnceCell lazily initialize the DB/crypto key.
+    // async_once_cell also has a `Lazy` type, which is closer to what we really want, but it seems
+    // to require spelling out the concrete type that implement Future, which would be awkward.
+    db_cell: OnceCell<CallbackResult<Mutex<UserDatabase>>>,
+    crypto_key_cell: OnceCell<CallbackResult<String>>,
+}
+
+
+#[uniffi::export]
+impl UserStore {
+    // Constructors are not yet implemented with proc-macros, but let's pretend they are
+    #[constructor]
+    pub fn new(logger: Box<dyn Logger>, key_store: Box<dyn Logger>, listener: Box<dyn EventListener>, queue: ThreadQueue) -> Self {
+        Self {
+            logger,
+            key_store,
+            listener,
+            queue,
+            db_cell: OnceCell::new(),
+            crypto_key_cell: OnceCell::new()
+        }
+    }
+
+    async fn get_db(&self) -> Result<&Mutex<UserDatabase>> {
+        Ok(db_cell.get_or_init(self.queue.schedule(|| Mutex::new(UserDatabase::new()))).await?)
+    }
+
+    async fn get_crypto_key(&self) -> Result<&str> {
+        Ok(db_cell.get_or_init(self.key_store.get()).await?)
+    }
+
+    /// Warmup by loading the database
+    pub async fn warmup(&self) {
+        self.get_db().await;
+    }
+
+    // Wrapper for a database operation that awaits on DB initialization and loading the crypto
+    // key.  f should be a synchronous function.
+    //
+    // This is the main point of this example.  This wrapper uses async callback interfaces to:
+    //   - Initialize the database and load the crypto key asynchronously
+    //   - Schedule the database operation in a background thread
+    //   - Log any errors
+    //   - Notify the application that the operation completed
```

Comment:
This feels like a nice version of [decorators](https://github.com/mozilla/uniffi-rs/pull/1051) to me.